Add Mobile Charger Connect

DerAndereAndi opened the pull request:
- The charger is or will be used by Audi, Bentley, Porsche
- Support implemented for
  - Login and auth token refresh
  - Status
  - Enabled
  - MaxCurrent
  - CurrentPower
  - ChargedEnergy
  - ChargingTime
- Enable is not supported as there is no known API to provide it
- Tested with 1 Phase standard home connector (230V)
- Test cases are not yet provided

DerAndereAndi committed:
Initial support for Mobile Charger Connect

- The charger is or will be used by Audi, Bentley, Porsche
- Support implemented for
  - Login and auth token refresh
  - Status
  - Enabled
  - MaxCurrent
  - CurrentPower
  - ChargedEnergy
  - ChargingTime
- Enable is not supported as there is no known API to provide it
- Tested with 1 Phase standard home connector (230V)
- Test cases are not yet provided

TheRealKerni commented:
Hi @andig 

this is my first try on go, so please don't be too harsh, but any feedback is more than welcome :)

I am planning on adding some tests as well, but wanted to provide a first try as soon as possible.

Thanks,
Andreas

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+
+// MCCTokenResponse is the apiLogin response
+type MCCTokenResponse struct {
+	AccessToken string `json:"token"`
```
Ich hab die Felder immer wie im JSON benannt, muss aber nicht sein

TheRealKerni replied:
Stimmt, der einzige denke ich wo ich es anders habe. Muss auch nicht sein :) Änder ich.

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+
+// MCCCurrentSession is the apiCurrentSession response
+type MCCCurrentSession struct {
+	Account           string     `json:"account"`
```
die JSON Tags brauchst Du zum lesen nicht solange das Feld wie das Tag heißt, ist Case Insensitive. Nur für's schreiben ist es für die Groß/Kleinschreibung dann wichtig. Stört aber auch nciht.

TheRealKerni replied:
Yep, das ist klar. Für mich war es für die Klarheit doch irgendwie schöner. Aber muss nicht sein.

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+	Soc               float64    `json:"soc"`
+	SolarEnergyShare  float64    `json:"solarEnergyShare"`
+	StartSoc          float64    `json:"startSoc"`
+	StartTime         *time.Time `json:"startTime"`
```
Time als Tristate?

TheRealKerni replied:
Der struct bildet 1:1 ab was von der API kommt. Ich habe da erst mal nicht geändert

andig replied:
Wieso das sternchen vor time? Das kann nil, true, false. Reicht nicht true false?

TheRealKerni replied:
Stimmt. Aber ich streich alle Element aus den structs raus die nicht verwendet werden. Was nicht verwendet wird, irritiert schon nicht

andig replied:
* musst du halt immer auf nil checken, ein bool nicht.

andig commented on charger/config.go:
```diff
@@ -27,6 +27,8 @@ func NewFromConfig(log *api.Logger, typ string, other map[string]interface{}) ap
 	case "simpleevse", "evse":
 		c = NewSimpleEVSEFromConfig(log, other)
+	case "porsche", "audi", "bentley", "mcc":
+		c = NewMCCFromConfig(log, other)
```
Ich würde das MCC ausschreiben für Wiedererkennung- lange Namen stören hier nicht unbedingt. Vielleicht als "MobileConnect"?

andig replied:
...von aussen sieht das dann nach idiomatischem go aus: `charger.NewMobileConnectFromConfig`

TheRealKerni replied:
Hatte ich Anfangs genau so, aber die anderen sind so kurz, da dachte ich ... :)

Also mache ich.

TheRealKerni replied:
Würdest du dann auch jedes Vorkommen "MCC" durch "MobileConnect" ersetzen? Fände ich dann "schöner"

andig replied:
Nicht in der methodensignatur bei *mcc. Da würde ich sogar *c nehmen wie alle anderen charger auch. Aber egal, *mcc geht auch.

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+func NewMCC(uri string, password string) *MCC {
+	// ignore the self signed certificate
+	customTransport := http.DefaultTransport.(*http.Transport).Clone()
+	customTransport.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
```
Wofür eigentlich? Ist die Box self-signed?

TheRealKerni replied:
Ja :(

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+		password:   password,
+	}
+
+	c.HTTPHelper.Log.WARN.Println("-- experimental --")
```
Hast Du doch getestet ;)

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+	client := &http.Client{Transport: customTransport}
+
+	c := &MCC{
+		HTTPHelper: api.NewHTTPHelperWithClient(api.NewLogger("mcc"), client),
```
bitte "mcc " damit alle 4 zeichen lang sind

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+
+	c := &MCC{
+		HTTPHelper: api.NewHTTPHelperWithClient(api.NewLogger("mcc"), client),
+		uri:        uri,
```
uri mit StringTrimRight("/") normalisieren

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+
+// construct the URL for a given apiFunction
+func (mcc *MCC) apiURL(api apiFunction) string {
+	return fmt.Sprintf("%s%s", mcc.uri, api)
```
analog %s/%s zu den anderen APIs

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+			}
+
+			return fmt.Errorf("response: %s", error.Error)
+
```
leerzeile weg

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+
+		mcc.token = tr.AccessToken
+		// According to the Web Interface, the token is valid for 10 minutes
+		mcc.tokenValid = time.Now().Add(time.Duration(10) * time.Minute)
```
das time.Duration brauchst Du nicht (glaub ich...)

andig replied:
Oder doch?

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+		// the web interface updates the token every 2 minutes, so lets do the same here
+		mcc.tokenRefresh = time.Now().Add(time.Duration(2) * time.Minute)
+
+		return nil
```
kann weg, ist ja hier nil

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+	uri := fmt.Sprintf("%s%s", mcc.uri, apiLogin)
+
+	data := url.Values{
+		"user": []string{"user"},
```
"user"??? hast du wirklich getestet?

TheRealKerni replied:
Yep. Natürlich :) es gibt 2 user (Heimanwender und Kundendienst), die kann man nicht ändern, im Web Interface ist das ein "Pulldown" zur Auswahl.

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	err = mcc.fetchToken(req)
```
return mcc.fetchToken...

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+
+	req.Header.Set("Authorization", fmt.Sprintf(" Bearer %s", mcc.token))
+
+	err = mcc.fetchToken(req)
```
dito

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+		return req, err
+	}
+
+	req.Header.Set("Authorization", fmt.Sprintf(" Bearer %s", mcc.token))
```
space for Bearer ist sehr ungewöhnlich?

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+
+// creates a custom http request that contains the auth token
+func (mcc *MCC) customRequest(method, uri string) (*http.Request, error) {
+	// do we need to login?
```
methode sollte request heissen analog der anderen APIs

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+		var result string
+		err = json.Unmarshal(b, &result)
+
+		if err == nil {
```
if err == nil && result...

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+	b, err := mcc.Request(req)
+	if err == nil {
+		var unquote string
+		if err := json.Unmarshal([]byte(b), &unquote); err != nil {
```
wofür brauchst Du das doppelte unquote?

TheRealKerni replied:
Weil die API double encoded JSON zurückschickt 😢 

Beispiel:
```
"{\"carCable\":5,\"gridCable\":8,\"hwfpMaxLimit\":32,\"maxValue\":10,\"minValue\":6,\"value\":10}"
```

andig replied:
Einfacher: https://stackoverflow.com/questions/16846553/how-to-unmarshal-an-escaped-json-string-in-go#16847249

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+func (mcc *MCC) Status() (api.ChargeStatus, error) {
+	var chargeState int64
+
+	err := mcc.getValue(mcc.apiURL(apiChargeState), &chargeState)
```
if err:=...; err!=nil {

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+		return api.StatusNone, err
+	}
+
+	// Charge State values and mapping to ChargeStatus
```
würde ich unten an die cases mit ranschreiben- weniger zeilen ;)

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+func (mcc *MCC) Enable(enable bool) error {
+	// As we don't know of the API to disable charging this for now always returns an error
+
+	return fmt.Errorf("The device doesn't allow to enable/disable the device via an API")
```
dann nil zurück geben, sonst gehts nicht!

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+	// Check if the car is connected and Paused, Active, or Finished
+	var chargeState int64
+
+	err := mcc.getValue(mcc.apiURL(apiChargeState), &chargeState)
```
einzeiler?

andig replied:
aber generell merkwürdig- unmarshal json in einen int? Ist das wirklich json und wie sieht das aus?

TheRealKerni replied:
Ne, das ist nun mal wirklich Blödsinn 😁 Da kommt einfach nur ein "String" zurück.

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+	// and then return an error if the value is outside of the limits or
+	// otherwise set the new value
+	var cableInformation MCCCurrentCableInformation
+	err := mcc.getEscapedJSON(mcc.apiURL(apiCurrentCableInformation), &cableInformation)
```
das muss aber nur einmal passieren: schau mal sync/once und once.Do an

TheRealKerni replied:
Schaue mir das gerade an, aber so glücklich macht mich das nicht. Falls beim Aufruf ein Fehler passiert ist es ziemlich umständlich das dann erneut versuchen zu lassen. 
Und danach muss ich auch immer testen ob `maxValue` einen sinnvollen Wert hat um nicht davor schon abzubrechen. 
Dazu noch `cableInformation` hoch ins struct ziehen. Ob es das Wert ist?

andig replied:
Du klnntest einfach testen ob max schon einen wert hat und falls nicht aufrufen?

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+	}
+
+	url := fmt.Sprintf("%s%d", mcc.apiURL(apiSetCurrentLimit), current)
+	fmt.Println(url)
```
leftover

TheRealKerni replied:
Ups 🤦

andig replied:
Du musst auch nicht kommentieren- wenns in neuem commit geändert ist mach gh die commants automatisch zu ;)

TheRealKerni replied:
Ich kommuniziere nun mal gerne 😁

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+	url := fmt.Sprintf("%s%d", mcc.apiURL(apiSetCurrentLimit), current)
+	fmt.Println(url)
+
+	if err := mcc.putValue(url); err != nil {
```
return ...

TheRealKerni replied:
Fehlende Leerzeile oder was meinst?

andig replied:
Do kannst direkt die funktion returnen, da passiert ja zwischendurch nix mehr

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+// CurrentPower implements the Meter interface.
+func (mcc *MCC) CurrentPower() (float64, error) {
+	var energy MCCEnergy
+	err := mcc.getEscapedJSON(mcc.apiURL(apiEnergy), &energy)
```
return ...

TheRealKerni replied:
Ich denke ich bin blind, oder da geht kein return

andig replied:
Ups.

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,406 @@
+// ChargingTime yields current charge run duration
+func (mcc *MCC) ChargingTime() (time.Duration, error) {
+	var currentSession MCCCurrentSession
+	err := mcc.getEscapedJSON(mcc.apiURL(apiCurrentSession), &currentSession)
```
return...

TheRealKerni replied:
hier ebenso blind

andig commented on evcc.dist.yaml:
```diff
@@ -83,6 +83,10 @@ chargers:
+- name: mcc
+  type: mcc # Mobile Charger Connect (Audi, Bentley, Porsche)
+  uri: https://192.168.1.4 # Mobile Charger Connect address
+  password: # home user password
```
fehlt noch ein Eintrag in der Readme ;)

TheRealKerni replied:
Stimmt 😄 Aber ist ja WIP, da MUSS dann schon noch was kommen...

andig replied:
😂

andig reviewed (changes requested):
Stark für den erste Go Code!

andig commented:
Du kannst einfach neue commits nachschieben.

DerAndereAndi committed:
Tons of changes due discussion

Not all, but getting there

andig commented on charger/mcc.go:
```diff
@@ -29,66 +29,36 @@ type MCCErrorResponse struct {
-	L1 MCCEnergyPhase `json:"L1"`
-	L2 MCCEnergyPhase `json:"L2"`
-	L3 MCCEnergyPhase `json:"L3"`
+	L1 MCCEnergyPhase
```
Kannst du auch in 1 zeile mit , schreiben

andig commented on charger/mcc.go:
```diff
@@ -29,66 +29,36 @@ type MCCErrorResponse struct {
-	MaxValue     int64 `json:"maxValue"`
-	MinValue     int64 `json:"minValue"`
-	Value        int64 `json:"value"`
+	MaxValue int64
```
Dito

andig commented on charger/mcc.go:
```diff
@@ -97,70 +67,65 @@ type MCC struct {
-		HTTPHelper: api.NewHTTPHelperWithClient(api.NewLogger("mcc"), client),
-		uri:        uri,
+	mcc := &MobileConnect{
+		HTTPHelper: api.NewHTTPHelperWithClient(api.NewLogger("MobileConnect"), client),
```
Ligger bitte „mcc „

andig commented on charger/mcc.go:
```diff
@@ -282,50 +241,44 @@ func (mcc *MCC) getEscapedJSON(uri string, result interface{}) error {
-	err := mcc.getValue(mcc.apiURL(apiChargeState), &chargeState)
-
-	if err != nil {
+	if err := mcc.getValue(mcc.apiURL(apiChargeState), &chargeState); err != nil {
```
Dito- int64 als json? Wozu?

andig commented on charger/mcc.go:
```diff
@@ -366,25 +318,20 @@ func (mcc *MCC) MaxCurrent(current int64) error {
+func (mcc *MobileConnect) ChargedEnergy() (float64, error) {
 	var currentSession MCCCurrentSession
 	err := mcc.getEscapedJSON(mcc.apiURL(apiCurrentSession), &currentSession)
 	if err != nil {
```
Kannst du auch zusammen fassen mit dem return.

andig commented on charger/mcc.go:
```diff
@@ -395,7 +342,7 @@ func (mcc *MCC) ChargedEnergy() (float64, error) {
+func (mcc *MobileConnect) ChargingTime() (time.Duration, error) {
 	var currentSession MCCCurrentSession
 	err := mcc.getEscapedJSON(mcc.apiURL(apiCurrentSession), &currentSession)
 	if err != nil {
```
Dito, dazu unten die duration?

andig commented on charger/config.go:
```diff
@@ -27,6 +27,8 @@ func NewFromConfig(log *api.Logger, typ string, other map[string]interface{}) ap
 	case "simpleevse", "evse":
 		c = NewSimpleEVSEFromConfig(log, other)
+	case "porsche", "audi", "bentley", "mcc":
+		c = NewMobileConnectFromConfig(log, other)
```
Wäre nicht evtl mcc oder mobileconnect hier besser? Not sure :/

TheRealKerni replied:
Bin ich recht leidenschaftslos. Sag du :)

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,353 @@
+	b, err := mcc.Request(req)
+	if err == nil {
+		var result string
+		err = json.Unmarshal(b, &result)
```
Wofür json?

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,353 @@
+		return err
+	}
+
+	b, err := mcc.Request(req)
```
Warum nicht RequestJSON?

DerAndereAndi committed:
More cleanup

andig commented on api/http.go:
```diff
@@ -22,6 +22,16 @@ func NewHTTPHelper(log *Logger) *HTTPHelper {
 }
 
+// NewHTTPHelperWithClient creates http helper for simplified PUT GET logic
+// This variant requires a client to be passed over which allows it to have custom settings
```
Brauchst du nicht- Client ist public und kann geändert werden

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,352 @@
+	client := &http.Client{Transport: customTransport}
+
+	mcc := &MobileConnect{
+		HTTPHelper: api.NewHTTPHelperWithClient(api.NewLogger("mcc "), client),
```
Siehe oben, geht auch ohne die neue methode

DerAndereAndi committed:
Updated readme

Added references to Mobile Charger Connect support

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,352 @@
+}
+
+// use http PUT to set a value on the URI, the value should be URL encoded in the URI parameter
+func (mcc *MobileConnect) putValue(uri string) ([]byte, error) {
```
Brauchst du nur 1x? Würde such die methode verzichten

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,352 @@
+}
+
+// use http GET to fetch a non structured value from an URI and stores it in result
+func (mcc *MobileConnect) getValue(uri string) ([]byte, error) {
```
Dito?

TheRealKerni replied:
Die wird in `Status()` und `Enabled()` benutzt

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,352 @@
+
+	b, err := mcc.Request(req)
+	if err == nil {
+		s, _ := strconv.Unquote(strings.Trim(string(b), "\n"))
```
Ich würd den error check trotzdem drin lassen

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,352 @@
+		return err
+	}
+
+	// return value is returned in the format "OK"\n
```
Kann kombiniert werden?

TheRealKerni replied:
Was meinst du?

andig replied:
Trim nimmt ein cutset: https://golang.org/pkg/strings/#Trim, also mehrere zeichen 👍🏻

DerAndereAndi committed:
Use the public Client instead of creating new one

DerAndereAndi committed:
Remove put function as it is only used once

DerAndereAndi committed:
Add another error check

TheRealKerni commented on charger/mcc.go:
```diff
@@ -0,0 +1,350 @@
+// ChargedEnergy implements the ChargeRater interface.
+func (mcc *MobileConnect) ChargedEnergy() (float64, error) {
+	var currentSession MCCCurrentSession
+	if err := mcc.getEscapedJSON(mcc.apiURL(apiCurrentSession), &currentSession); err != nil {
```
Habe gerade herausgefunden, dass wenn der Charger gestartet wurde und damit noch keine Session gestartet wurde und das Auto auch nicht verbunden wurde, dann kommt kein JSON und auch kein Escaped JSON, sondern einfach nur `""` zurück. Der Code würde dann momentan `0` zurück liefern aber auch einen error mit `unexpected end of JSON input`.

Gleiches dann in `ChargingTime()`. 

Sollte da das Verhalten geändert werden?

andig replied:
Wie wär statusa als default? Chargingtime wäre dann ja 0. das api ist schrott...

TheRealKerni replied:
"schrott" ist aber noch nett formuliert 😁 

`Status()` liefert `StatusA` in dem Fall.
 
Apropo: Diese `statusa,...` gefallen mir nicht so wirklich. Könnte man das nicht etwas "schöner" machen? :)

andig replied:
Hast Du eine Idee (per Mail)?

DerAndereAndi committed:
A little simplification

TheRealKerni commented:
Ich würde gerne für die super genial schönen API Aufrufe Tests einbauen. Auch vor allem weil ich damit rechne dass in neuen Software Versionen an denen schon mal gedreht werden wird.

Hast du da einen Hinweis wie man das hier am Besten anstellen könnte?

andig commented:
Google mal nach golang mock http. Im Prinzip würde mal wohl den Client durch ein Doer Interface ersetzen. Du könntest aber auch den ganzen HTTP Helper mocken (siehe api.go) und da einfach den Mock durch fake Antworten ersetzen. Mocks siehe Tests in core/wrapper.

Viele Grüße, Andreas


> On 18. Apr 2020, at 12:43, Andreas Linde <notifications@github.com> wrote:
> 
> 
> Ich würde gerne für die super genial schönen API Aufrufe Tests einbauen. Auch vor allem weil ich damit rechne dass in neuen Software Versionen an denen schon mal gedreht werden wird.
> 
> Hast du da einen Hinweis wie man das hier am Besten anstellen könnte?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/andig/evcc/pull/40#issuecomment-615839717>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AABND34E2O4S3UAMKSV2OC3RNF74PANCNFSM4MK4GTLA>.
>

DerAndereAndi committed:
First try at adding tests

- Followed approach at http://hassansin.github.io/Unit-Testing-http-client-in-Go
- Not happy yet by providing the HTTP response body as escaped strings in code
- The base test structure was generated with VS Code automated test file generation

TheRealKerni commented:
So, habe etwas gebaut und Tests hinzugefügt. Bin noch nicht ganz glücklich aber auf Feedback gespannt 😄

andig commented on api/http.go:
```diff
@@ -54,6 +54,13 @@ func (r *HTTPHelper) decodeJSON(resp *http.Response, err error, res interface{})
 }
 
+// Request executes HTTP request returns the response body
+// This variant uses the http method set on the request
```
can we drop this line? doesn't add any information ;)

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,347 @@
+	customTransport := http.DefaultTransport.(*http.Transport).Clone()
+	customTransport.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+
+	mcc.HTTPHelper.Client.Transport = customTransport
```
dumme idee: warum nicht 

    mcc.HTTPHelper.Client.Transport.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}

TheRealKerni replied:
Weil `mcc.HTTPHelper.Client.Transport.TLSClientConfig undefined (type http.RoundTripper has no field or method TLSClientConfig)`

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,347 @@
+	return fmt.Sprintf("%s/%s", mcc.uri, api)
+}
+
+// proces the http request to fetch the auth token for a login or refresh request
```
process

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,347 @@
+
+	b, err := mcc.Request(req)
+	if err == nil {
+		s, err := strconv.Unquote(strings.Trim(string(b), "\n"))
```
einfacher: 

    if s,err:= ...; err == nil {
      err = json.Unmarshal([]byte(s), &result)
    }

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,347 @@
+
+	switch chargeState {
+	case 0:
+		// 0: Unplugged		StatusA
```
ich würde die kommentare noch jeweils hinters case schreiben und status weglassen

andig commented on charger/mcc.go:
```diff
@@ -0,0 +1,347 @@
+
+	// return value is returned in the format "OK"\n
+	if strings.Trim(string(b), "\n\"") != "OK" {
+		return fmt.Errorf("Call returned an unexpected error")
```
Das failed beim linter: errorf ohne formatierung. Gib doch noch %s mit aus damit man den Wert Sieht. Alle Fehlermeldung: 

    maxcurrent unexpected response: %s

andig commented on charger/mcc_test.go:
```diff
@@ -0,0 +1,479 @@
+	"github.com/andig/evcc/api"
+)
+
+// RoundTripFunc .
```
Lowecase, dann musst du sie nicht "fake" kommentieren

andig commented on charger/mcc_test.go:
```diff
@@ -0,0 +1,479 @@
+type RoundTripFunc func(req *http.Request) *http.Response
+
+// RoundTrip .
+func (f RoundTripFunc) RoundTrip(req *http.Request) (*http.Response, error) {
```
nice

TheRealKerni replied:
Wenn es nur von mir wäre :) Ich würde da im Code noch gerne die Referenz reinschreiben wo das herkommt, ok für dich?

andig replied:
Klar, amch ich auch immer

andig commented on charger/mcc_test.go:
```diff
@@ -0,0 +1,479 @@
+		{"login - wrong password", []apiResponse{{apiLogin, "{\n    \"error\": \"wrong password\"\n}"}}, "wrong", true},
+		{"login - bad return", []apiResponse{{apiLogin, "{{\n    \"error\": \"wrong password\"\n}"}}, "wrong", true},
+	}
+	for _, tt := range tests {
```
ich hab die test cases sonst immer `tc` genannt

andig commented on charger/mcc_test.go:
```diff
@@ -0,0 +1,479 @@
+			}
+		})
+	}
+}
```
sieht gut ausm, ist aber *richtig* viel aufwand für eine wenig verbreitete box ;)

TheRealKerni replied:
Warte mal bis die elektrischen Audi Firmenwagen mehr kommen ... 😁

andig commented:
Erledigt- die Tests finde ich echt gut!

Viele Grüße, Andreas


> On 18. Apr 2020, at 18:34, Andreas Linde <notifications@github.com> wrote:
> 
> 
> So, habe etwas gebaut und Tests hinzugefügt. Bin noch nicht ganz glücklich aber auf Feedback gespannt 😄
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/andig/evcc/pull/40#issuecomment-615899932>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AABND3ZOU36Y7HRIHBXBSZDRNHJBRANCNFSM4MK4GTLA>.
>

DerAndereAndi committed:
Fix charge states

Fix states with the explanation from https://evsim.gonium.net

StatusA		Unplugged
StatusB		Plugged
StatusC		Charging
StatusD		Charging with ventilator
StatusE		Error (Kurzschluss)
StatusF		Error (Ausfall Wallbox)

DerAndereAndi committed:
Various changes

- Fixed typo
- Simplified code
- Aligned style to codebase
- Remove non-implemented test cases

DerAndereAndi committed:
Added reference

Added a reference to the source for the http test approach

DerAndereAndi committed:
Fixed another typo

TheRealKerni commented:
If this looks good, I am ready :)

DerAndereAndi committed:
More ChargeStatus fixes

This is confusing :(

DerAndereAndi committed:
Minor typos

DerAndereAndi committed:
Fixed tests

TheRealKerni commented:
Wir sollten hier noch die Erkenntnisse aus https://github.com/andig/evcc/pull/42 abwarten, denn momentan ist der `ChargeStatus` wohl nicht so gesetzt wie es benötigt wird,

DerAndereAndi committed:
Fix test case software version reference

The test cases provided correlate to the wallbox software version 2914

DerAndereAndi committed:
Update charge state with new understanding

TheRealKerni commented:
So, das sollte es dann sein. 😄
